File: haskell/src/HexagramData.hs
{-# LANGUAGE OverloadedStrings #-}

{-|
Module      : HexagramData
Description : Traditional I Ching hexagram names and judgments (King Wen sequence)
Copyright   : (c) WhiteMagic Contributors, 2025
License     : MIT

All 64 hexagrams with their traditional Chinese names, pinyin, and judgment
text. Data sourced from the classical King Wen ordering used in standard
I Ching translations.
-}

module HexagramData
    ( HexagramInfo(..)
    , getHexagramInfo
    ) where

import Data.Text (Text)
import qualified Data.Map.Strict as Map

-- | Metadata for a single hexagram
data HexagramInfo = HexagramInfo
    { hexName    :: Text   -- ^ Traditional Chinese name
    , hexPinyin  :: Text   -- ^ Pinyin romanization
    , judgment   :: Text   -- ^ Classical judgment / core teaching
    } deriving (Show, Eq)

-- | Lookup hexagram info by King Wen number (1-64)
getHexagramInfo :: Int -> Maybe HexagramInfo
getHexagramInfo n = Map.lookup n hexagramTable

-- | Full table: King Wen number -> HexagramInfo
hexagramTable :: Map.Map Int HexagramInfo
hexagramTable = Map.fromList
    [ ( 1, HexagramInfo "乾" "Qian"  "The Creative. Heaven above heaven. Great success through perseverance and inner truth.")
    , ( 2, HexagramInfo "坤" "Kun"   "The Receptive. Earth above earth. Devoted service and quiet strength bring success.")
    , ( 3, HexagramInfo "屯" "Zhun"  "Difficulty at the Beginning. Water above thunder. Perseverance through initial hardship yields eventual success.")
    , ( 4, HexagramInfo "蒙" "Meng"  "Youthful Folly. Mountain above water. Accept instruction with humility; the teacher finds the pupil.")
    , ( 5, HexagramInfo "需" "Xu"    "Waiting. Water above heaven. Patience and readiness; do not force action.")
    , ( 6, HexagramInfo "訟" "Song"  "Conflict. Heaven above water. Avoid prolonged dispute; seek compromise before engaging others.")
    , ( 7, HexagramInfo "師" "Shi"   "The Army. Earth above water. Leadership and discipline bring collective success.")
    , ( 8, HexagramInfo "比" "Bi"    "Holding Together. Water above earth. Unity and mutual support create lasting bonds.")
    , ( 9, HexagramInfo "小畜" "Xiao Xu" "The Power of the Small. Wind above heaven. Gentle persistence accumulates quiet influence.")
    , (10, HexagramInfo "履" "Lu"    "Treading. Heaven above lake. Walk carefully and with awareness; respect boundaries.")
    , (11, HexagramInfo "泰" "Tai"   "Peace. Earth above heaven. Harmony between heaven and earth; the right moment for action.")
    , (12, HexagramInfo "否" "Pi"    "Standstill. Heaven above earth. Stagnation and disconnection; withdraw and conserve energy.")
    , (13, HexagramInfo "同人" "Tong Ren" "Fellowship of Men. Heaven above fire. Harmonious cooperation with others toward shared purpose.")
    , (14, HexagramInfo "大畜" "Da Xu" "Great Accumulation. Mountain above heaven. Vast resources gathered through patience and merit.")
    , (15, HexagramInfo "謙" "Qian"  "Modesty. Earth above mountain. Humility is the highest virtue; it brings lasting respect.")
    , (16, HexagramInfo "豫" "Yu"    "Enthusiasm. Thunder above earth. Joyful preparation; inspire others through genuine delight.")
    , (17, HexagramInfo "隨" "Sui"   "Following. Lake above thunder. Adapt to circumstances with open awareness; yielding is strength.")
    , (18, HexagramInfo "蠱" "Gu"    "Work on the Decayed. Mountain above wind. Address what has decayed; renewal requires honest effort.")
    , (19, HexagramInfo "臨" "Lin"   "Approach. Earth above lake. The time of coming and engagement; welcome opportunity with openness.")
    , (20, HexagramInfo "觀" "Guan"  "Contemplation. Wind above earth. Observe carefully before acting; perception precedes understanding.")
    , (21, HexagramInfo "噬嗑" "She He" "Biting Through. Fire above thunder. Resolve obstacles decisively; clarity comes through direct action.")
    , (22, HexagramInfo "賁" "Bi"    "Grace. Fire above mountain. Beauty and refinement add meaning; outer form reflects inner truth.")
    , (23, HexagramInfo "剥" "Bo"    "Splitting Apart. Mountain above earth. Decline is natural; do not resist what must fall away.")
    , (24, HexagramInfo "復" "Fu"    "Return. Earth above thunder. The cycle turns; light returns after darkness. Rest and renew.")
    , (25, HexagramInfo "無妨" "Wu Wang" "Innocence. Heaven above thunder. Act from pure intention; sincerity brings unexpected fortune.")
    , (26, HexagramInfo "大畜" "Da Xu" "Great Accumulation. Mountain above heaven. Contain and nourish power; the mountain holds the sky.")
    , (27, HexagramInfo "頌" "Yi"    "Nourishment. Mountain above thunder. Attend to what you feed: body, mind, and others.")
    , (28, HexagramInfo "大過" "Da Guo" "Preponderance of the Great. Lake above wind. The weight of the situation demands courage and decisiveness.")
    , (29, HexagramInfo "坎" "Kan"   "The Abysmal. Water above water. Navigate danger with steadiness and inner light; do not panic.")
    , (30, HexagramInfo "離" "Li"    "The Clinging. Fire above fire. Clarity and awareness illuminate the path; cling to truth.")
    , (31, HexagramInfo "咸" "Xian"  "Influence. Lake above mountain. Mutual attraction and sensitivity; receptivity creates connection.")
    , (32, HexagramInfo "恒" "Heng"  "Duration. Wind above thunder. Endurance and consistency; commit fully to what you begin.")
    , (33, HexagramInfo "遁" "Dun"   "Retreat. Heaven above mountain. Strategic withdrawal preserves strength; know when to step back.")
    , (34, HexagramInfo "大壯" "Da Zhuang" "The Power of the Great. Thunder above heaven. Great strength must be tempered with wisdom.")
    , (35, HexagramInfo "進" "Jin"   "Progress. Fire above earth. Steady advancement into the light; conditions favor forward movement.")
    , (36, HexagramInfo "明夷" "Ming Yi" "Darkening of the Light. Earth above fire. Protect your inner light during difficult times; maintain integrity.")
    , (37, HexagramInfo "家人" "Jia Ren" "The Family. Wind above fire. Harmonious roles within family create lasting stability.")
    , (38, HexagramInfo "睽" "Kui"   "Opposition. Fire above lake. Differences can be bridged; understanding lies beneath apparent conflict.")
    , (39, HexagramInfo "蹇" "Jian"  "Obstruction. Water above mountain. Difficulty ahead; seek guidance and do not travel alone.")
    , (40, HexagramInfo "解" "Jie"   "Deliverance. Thunder above water. Release and liberation; forgive and move forward freely.")
    , (41, HexagramInfo "損" "Sun"   "Decrease. Mountain above lake. Let go of excess; simplification brings clarity and renewal.")
    , (42, HexagramInfo "益" "Yi"    "Increase. Wind above thunder. The time for generous giving and mutual benefit has arrived.")
    , (43, HexagramInfo "夬" "Gua"   "Breakthrough. Lake above heaven. Decisive action resolves long-standing tension; speak truth.")
    , (44, HexagramInfo "姤" "Gou"   "Coming Together. Wind above heaven. Unexpected encounter; be discerning about what you welcome.")
    , (45, HexagramInfo "萃" "Cui"   "Gathering. Lake above earth. Assemble resources and allies; collective strength serves the cause.")
    , (46, HexagramInfo "升" "Sheng" "Pushing Upward. Wind above earth. Gradual rise through steady effort; the time is right.")
    , (47, HexagramInfo "困" "Kun"   "Oppression. Lake above water. Exhaustion and constraint; maintain inner peace through difficulty.")
    , (48, HexagramInfo "井" "Jing"  "The Well. Water above wind. Nourishment is always available; draw from the deep source.")
    , (49, HexagramInfo "革" "Ge"    "Revolution. Lake above fire. Transformation is needed; change comes when old forms can no longer serve.")
    , (50, HexagramInfo "鼎" "Ding"  "The Cauldron. Fire above wind. The vessel of civilization; cooking and transformation nourish all.")
    , (51, HexagramInfo "震" "Zhen"  "The Arousing. Thunder above thunder. Shock awakens; after the storm, clarity and fresh beginning.")
    , (52, HexagramInfo "艮" "Gen"   "Keeping Still. Mountain above mountain. Rest and meditation; stillness reveals what motion obscures.")
    , (53, HexagramInfo "漸" "Jian"  "Development. Wind above mountain. Gradual progress like a tree growing; patience with the process.")
    , (54, HexagramInfo "歸妹" "Gui Mei" "The Marrying Maiden. Thunder above lake. Understand your true position; act from a place of clarity.")
    , (55, HexagramInfo "豐" "Feng"  "Abundance. Thunder above fire. Fullness and prosperity; enjoy the peak, knowing it is transient.")
    , (56, HexagramInfo "旅" "Lu"    "The Wanderer. Fire above mountain. Travel light and with respect; the stranger finds kindness in humility.")
    , (57, HexagramInfo "巽" "Xun"   "The Gentle. Wind above wind. Gentle penetration and influence; persistence without force achieves results.")
    , (58, HexagramInfo "兌" "Dui"   "The Joyous. Lake above lake. Joy shared is joy multiplied; openness and receptivity invite delight.")
    , (59, HexagramInfo "渙" "Huan"  "Dispersion. Wind above water. Dissolve what is rigid; flow and trust in the process of release.")
    , (60, HexagramInfo "節" "Jie"   "Limitation. Water above lake. Boundaries create structure; accept limits as the framework for growth.")
    , (61, HexagramInfo "中孚" "Zhong Fu" "Inner Truth. Wind above lake. Sincerity penetrates all barriers; truth spoken from the heart.")
    , (62, HexagramInfo "小過" "Xiao Guo" "Preponderance of the Small. Thunder above mountain. Small actions matter now; do not attempt the grand gesture.")
    , (63, HexagramInfo "既濟" "Ji Ji"  "After Completion. Water above fire. The task is done; maintain awareness, for disorder follows order.")
    , (64, HexagramInfo "未濟" "Wei Ji" "Before Completion. Fire above water. The task is not yet finished; perseverance will see it through.")
    ]

File: whitemagic/oracle/hexagram_data.py
"""Complete 64 I Ching Hexagrams - Traditional Interpretations

Binary perfection: 8 trigrams × 8 = 64 (matching DNA codons)
Each hexagram represents a unique pattern of change and wisdom.
"""

COMPLETE_HEXAGRAMS = [
    # 1-10
    (1, "The Creative", "乾 Qián", "The Creative works sublime success, furthering through perseverance",
     "Heaven's movement is full of power", "☰☰", "Continue with confidence and strength"),

    (2, "The Receptive", "坤 Kūn", "The Receptive brings supreme success through the perseverance of a mare",
     "Earth's disposition is receptive devotion", "☷☷", "Be receptive, nurturing, and yielding"),

    (3, "Difficulty at Beginning", "屯 Zhūn", "Difficulty at the beginning works supreme success through perseverance",
     "Clouds and thunder emerge", "☵☳", "Persist through chaos - order will emerge"),

    (4, "Youthful Folly", "蒙 Méng", "Inexperience seeks teaching; I do not seek the young fool, the young fool seeks me",
     "Spring at foot of mountain", "☶☵", "Embrace learning with humility"),

    (5, "Waiting", "需 Xū", "Waiting with sincerity brings success; perseverance brings good fortune",
     "Clouds rise to heaven", "☵☰", "Patient waiting with inner confidence"),

    (6, "Conflict", "訟 Sòng", "Conflict requires meeting halfway; going through brings danger",
     "Heaven and water move apart", "☰☵", "Seek compromise, avoid escalation"),

    (7, "The Army", "師 Shī", "The army needs experienced leadership; this brings good fortune",
     "Water beneath earth", "☷☵", "Discipline and order under wise command"),

    (8, "Holding Together", "比 Bǐ", "Union brings good fortune; inquire of the oracle again",
     "Water upon earth", "☵☷", "Build alliances with sincerity"),

    (9, "Small Taming", "小畜 Xiǎo Chù", "Gentle restraint brings success",
     "Wind moves across heaven", "☴☰", "Accumulate power through gentle persistence"),

    (10, "Treading", "履 Lǚ", "Treading carefully on the tiger's tail; it does not bite",
     "Heaven above, lake below", "☰☱", "Proceed with caution and respect"),

    # 11-20
    (11, "Peace", "泰 Tài", "Peace; the small departs, the great approaches",
     "Earth above, heaven below", "☷☰", "Harmony between heaven and earth"),

    (12, "Standstill", "否 Pǐ", "Standstill; the great departs, the small approaches",
     "Heaven above, earth below", "☰☷", "Withdraw and cultivate inner resources"),

    (13, "Fellowship", "同人 Tóng Rén", "Fellowship with others in the open brings success",
     "Heaven together with fire", "☰☲", "Unite with others for common purpose"),

    (14, "Great Possession", "大有 Dà Yǒu", "Possession in great measure; supreme success",
     "Fire above heaven", "☲☰", "Abundance through virtue and sharing"),

    (15, "Modesty", "謙 Qiān", "Modesty creates success; the superior person carries things through",
     "Earth above mountain", "☷☶", "Humility brings respect and advancement"),

    (16, "Enthusiasm", "豫 Yù", "Enthusiasm brings advantage; appoint helpers",
     "Thunder from earth", "☳☷", "Move forward with infectious joy"),

    (17, "Following", "隨 Suí", "Following has supreme success; perseverance furthers",
     "Thunder within lake", "☱☳", "Adapt and follow the time"),

    (18, "Work on the Decayed", "蠱 Gǔ", "Work on what has been spoiled brings supreme success",
     "Wind below mountain", "☶☴", "Repair corruption at its source"),

    (19, "Approach", "臨 Lín", "Approach brings supreme success; perseverance furthers",
     "Earth above lake", "☷☱", "Leadership through joyful accessibility"),

    (20, "Contemplation", "觀 Guān", "Contemplation; ablution made but not yet the offering",
     "Wind above earth", "☴☷", "Observe deeply before acting"),

    # 21-30
    (21, "Biting Through", "噬嗑 Shì Kè", "Biting through brings success; it is favorable to let justice be administered",
     "Fire and thunder", "☲☳", "Remove obstacles decisively"),

    (22, "Grace", "賁 Bì", "Grace brings success in small matters",
     "Fire below mountain", "☶☲", "Beauty and form enhance substance"),

    (23, "Splitting Apart", "剝 Bō", "Splitting apart; it does not further one to go anywhere",
     "Mountain above earth", "☶☷", "Yield when foundation crumbles"),

    (24, "Return", "復 Fù", "Return brings success; going out and coming in without error",
     "Thunder within earth", "☷☳", "Renewal after the turning point"),

    (25, "Innocence", "無妄 Wú Wàng", "Innocence brings supreme success; perseverance furthers",
     "Heaven beneath thunder", "☳☰", "Act from genuine spontaneity"),

    (26, "Great Taming", "大畜 Dà Chù", "Great taming brings perseverance; not eating at home brings good fortune",
     "Mountain above heaven", "☶☰", "Accumulate virtue and knowledge"),

    (27, "Nourishment", "頤 Yí", "Nourishment brings good fortune when seeking nourishment properly",
     "Mountain above thunder", "☶☳", "Be mindful of what you take in"),

    (28, "Great Excess", "大過 Dà Guò", "Preponderance of the great; the ridgepole sags",
     "Lake above wind", "☱☴", "Extraordinary times require extraordinary measures"),

    (29, "The Abysmal", "坎 Kǎn", "The Abysmal repeated brings danger but sincerity brings success",
     "Water doubled", "☵☵", "Flow through danger with constancy"),

    (30, "The Clinging", "離 Lí", "The Clinging; perseverance furthers; it brings success",
     "Fire doubled", "☲☲", "Cling to what is luminous and correct"),

    # 31-40
    (31, "Influence", "咸 Xián", "Influence brings success; perseverance furthers; taking a maiden brings good fortune",
     "Lake above mountain", "☱☶", "Mutual attraction and receptivity"),

    (32, "Duration", "恆 Héng", "Duration brings success without blame; perseverance furthers",
     "Thunder above wind", "☳☴", "Constancy and endurance prevail"),

    (33, "Retreat", "遯 Dùn", "Retreat brings success in small matters",
     "Heaven above mountain", "☰☶", "Strategic withdrawal preserves strength"),

    (34, "Great Power", "大壯 Dà Zhuàng", "Great power; perseverance brings good fortune",
     "Thunder above heaven", "☳☰", "Strength tempered by righteousness"),

    (35, "Progress", "晉 Jìn", "Progress like the sun rising over earth",
     "Fire above earth", "☲☷", "Advancement through clarity"),

    (36, "Darkening of Light", "明夷 Míng Yí", "Darkening of the light; perseverance in adversity furthers",
     "Earth above fire", "☷☲", "Hide brilliance in dark times"),

    (37, "The Family", "家人 Jiā Rén", "The family; perseverance of the woman brings good fortune",
     "Wind above fire", "☴☲", "Establish order within the home"),

    (38, "Opposition", "睽 Kuí", "Opposition; in small matters, good fortune",
     "Fire above lake", "☲☱", "Honor differences while seeking common ground"),

    (39, "Obstruction", "蹇 Jiǎn", "Obstruction; the southwest furthers, the northeast does not",
     "Water above mountain", "☵☶", "Face difficulty; seek help from allies"),

    (40, "Deliverance", "解 Xiè", "Deliverance; the southwest furthers; if nothing more, return brings good fortune",
     "Thunder above water", "☳☵", "Release from tension brings relief"),

    # 41-50
    (41, "Decrease", "損 Sǔn", "Decrease combined with sincerity brings supreme good fortune",
     "Mountain above lake", "☶☱", "Simplify and decrease to increase essence"),

    (42, "Increase", "益 Yì", "Increase; it furthers one to undertake something",
     "Wind above thunder", "☴☳", "Augment the good; share abundance"),

    (43, "Breakthrough", "夬 Guài", "Breakthrough; one must resolutely make the matter known",
     "Lake above heaven", "☱☰", "Decisive action ends the old"),

    (44, "Coming to Meet", "姤 Gòu", "Coming to meet; the maiden is powerful",
     "Heaven above wind", "☰☴", "Unexpected encounter requires caution"),

    (45, "Gathering Together", "萃 Cuì", "Gathering together brings success; the king approaches his temple",
     "Lake above earth", "☱☷", "Assemble forces for common purpose"),

    (46, "Pushing Upward", "升 Shēng", "Pushing upward brings supreme success; see the great person",
     "Earth above wind", "☷☴", "Steady growth through effort"),

    (47, "Oppression", "困 Kùn", "Oppression brings success to the persevering great person",
     "Lake above water", "☱☵", "Maintain integrity under constraint"),

    (48, "The Well", "井 Jǐng", "The well; the town may be changed but not the well",
     "Water above wind", "☵☴", "Inexhaustible source of nourishment"),

    (49, "Revolution", "革 Gé", "Revolution; supreme success through firm correctness",
     "Lake above fire", "☱☲", "Transform when the time is ripe"),

    (50, "The Caldron", "鼎 Dǐng", "The caldron brings supreme good fortune and success",
     "Fire above wind", "☲☴", "Nourishment and transformation through culture"),

    # 51-60
    (51, "The Arousing", "震 Zhèn", "The arousing brings success; thunder comes with shock",
     "Thunder doubled", "☳☳", "Shock awakens; maintain composure"),

    (52, "Keeping Still", "艮 Gèn", "Keeping still; meditate until movement ceases",
     "Mountain doubled", "☶☶", "Stillness and contemplation bring clarity"),

    (53, "Development", "漸 Jiàn", "Development; gradual progress brings good fortune",
     "Wind above mountain", "☴☶", "Patient development like tree growth"),

    (54, "The Marrying Maiden", "歸妹 Guī Mèi", "The marrying maiden; undertakings bring misfortune",
     "Thunder above lake", "☳☱", "Act with propriety in relationships"),

    (55, "Abundance", "豐 Fēng", "Abundance brings success; the king attains it",
     "Thunder above fire", "☳☲", "Fullness at its peak; prepare for decline"),

    (56, "The Wanderer", "旅 Lǚ", "The wanderer; success in small matters through perseverance",
     "Fire above mountain", "☲☶", "Travel light and remain adaptable"),

    (57, "The Gentle", "巽 Xùn", "The gentle penetrates; success in small matters",
     "Wind doubled", "☴☴", "Gentle penetration achieves results"),

    (58, "The Joyous", "兌 Duì", "The joyous brings success; perseverance is favorable",
     "Lake doubled", "☱☱", "Joy shared with others multiplies"),

    (59, "Dispersion", "渙 Huàn", "Dispersion brings success; the king approaches his temple",
     "Wind above water", "☴☵", "Dissolve rigidity and separateness"),

    (60, "Limitation", "節 Jié", "Limitation brings success; bitter limitation brings misfortune",
     "Water above lake", "☵☱", "Accept necessary limits with grace"),

    # 61-64
    (61, "Inner Truth", "中孚 Zhōng Fú", "Inner truth; it furthers one to cross the great water",
     "Wind above lake", "☴☱", "Sincerity penetrates to the core"),

    (62, "Small Exceeding", "小過 Xiǎo Guò", "Preponderance of the small brings success",
     "Thunder above mountain", "☳☶", "Attend to details; avoid grand ventures"),

    (63, "After Completion", "既濟 Jì Jì", "After completion brings success in small matters",
     "Water above fire", "☵☲", "Maintain vigilance at success"),

    (64, "Before Completion", "未濟 Wèi Jì", "Before completion brings success; careful perseverance furthers",
     "Fire above water", "☲☵", "Almost complete; take final steps with care"),
]
